collect_app/src/main/java/org/odk/collect/android/widgets/AnnotateWidget.java

```diff
@@ -99,32 +90,28 @@ public void onClick(View v) {
             }
         });
 
-        // finish complex layout
-        LinearLayout answerLayout = new LinearLayout(getContext());
-        answerLayout.setOrientation(LinearLayout.VERTICAL);
-
         answerLayout.addView(captureButton);
         answerLayout.addView(chooseButton);
         answerLayout.addView(annotateButton);
         answerLayout.addView(errorTextView);
 
         // and hide the capture, choose and annotate button if read-only
-        if (prompt.isReadOnly()) {
+        if (getFormEntryPrompt().isReadOnly()) {
             captureButton.setVisibility(View.GONE);
             chooseButton.setVisibility(View.GONE);
             annotateButton.setVisibility(View.GONE);
         }
         errorTextView.setVisibility(View.GONE);
+    }
 
-        // retrieve answer from data model and update ui
-        binaryName = prompt.getAnswerText();
-
+    @Override
```

Comment:
I’m working on an idea that won’t require you to override `setUpBinary` here. Overriding it duplicates code and complicates maintenance.

Reply:
I recommend you cherry pick this commit into your branch and see what you think of it.
https://github.com/dcbriccetti/collect/commit/28dce19a69401f917fe268bc6e911a4b165dc6b0

Reply:
After this is addressed, I will like the code, and suggest it be retested by QA.